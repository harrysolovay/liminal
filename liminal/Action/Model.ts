import type { Spec } from "../Spec.js"
import { ActionBase } from "./ActionBase.js"
import type { EventBase } from "./event_common.js"

export interface Model<S extends Spec = Spec> extends ActionBase<"Model", S> {
  key: string
  purpose: ModelPurpose
}

export type ModelPurpose = "language" | "embedding"

export function* Model<K extends string, P extends "language" | "embedding" = "language">(
  key: K,
  purpose: P = "language" as P,
): Generator<
  Model<{
    LanguageModel: P extends "language" ? K : never
    EmbeddingModel: P extends "embedding" ? K : never
    Event: ModelEvent<K, P>
  }>,
  void
> {
  return yield ActionBase("Model", { key, purpose })
}

export interface ModelEvent<K extends string = string, P extends ModelPurpose = ModelPurpose>
  extends EventBase<"Model"> {
  key: K
  purpose: P
}
