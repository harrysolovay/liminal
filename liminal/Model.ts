import type { StandardSchemaV1 } from "@standard-schema/spec"
import type { Action } from "./Action.ts"
import type { Agent } from "./Agent.ts"
import type { JSONObject } from "./util/JSONObject.ts"

export type Model = LanguageModel | EmbeddingModel

export interface LanguageModel extends ModelBase<"language"> {
  infer: (type: StandardSchemaV1<JSONObject, any> | undefined) => Agent<Action, any>
}

export interface EmbeddingModel extends ModelBase<"embedding"> {
  embed: (value: string) => Promise<Array<number>>
}

interface ModelBase<K extends ModelType> {
  type: K
}

export type ModelType = "language" | "embedding"
