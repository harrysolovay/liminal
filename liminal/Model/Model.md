# `Model`

Declare the intended use of a model by specifying a key to which the executor
ultimately binds a model.
