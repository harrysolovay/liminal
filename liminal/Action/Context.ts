import type { Action } from "./Action.js"
import type { ActorLike } from "../common/ActorLike.js"
import { ActionBase } from "./ActionBase.js"
import type { EventBase } from "./event_common.js"
import type { ActionEvent } from "./ActionEvent.js"
import type { Spec } from "../Spec.js"

export function Context<K extends string, Y extends Action, R = string>(
  key: K,
  system: string,
  implementation: ActorLike<Y, R>,
): Generator<
  Context<{
    Model: never
    Event: never
  }>,
  Awaited<R>
>
export function Context<K extends string, Y extends Action, R = string>(
  key: K,
  implementation: ActorLike<Y, R>,
): Generator<
  Context<{
    Model: never
    Event: never
  }>,
  Awaited<R>
>
export function* Context<Y extends Action, R = string>(
  key: string,
  a0: string | ActorLike,
  a1?: ActorLike,
): Generator<Context, Awaited<R>> {
  return yield ActionBase("Context", {
    key,
    system: typeof a0 === "string" ? a0 : undefined,
    implementation: typeof a0 === "string" ? a1 : a0,
  })
}

export interface Context<S extends Spec = Spec> extends ActionBase<"Context", S> {
  key: string
  system: string | undefined
  implementation?: ActorLike
}

export interface ContextEvent<K extends string = string, E extends ActionEvent = ActionEvent>
  extends EventBase<"Context"> {
  key: K
  inner: E
}
