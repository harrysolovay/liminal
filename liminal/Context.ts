import { AsyncLocalStorage } from "node:async_hooks"

const storage = new AsyncLocalStorage<Context>()

export class Context extends Map<ContextPart, unknown> {
  static get(): Context | undefined {
    return storage.getStore()
  }

  run<R>(f: () => R): R {
    return storage.run(this, f)
  }

  fork(): Context {
    const context = new Context()
    for (const [handle, value] of this.entries()) {
      if (!context.has(handle)) {
        context.set(handle, handle.fork(value))
      }
    }
    return context
  }
}

export interface ContextPart<V = any> {
  fork(parent?: V): V
  get(): V | undefined
  debug?: string
}

export function ContextPart<V>(fork: (parent?: V) => V, debug?: string): ContextPart<V> {
  const self: ContextPart<V> = {
    fork,
    get() {
      return Context.get()?.get(self) as never
    },
    ...debug && { debug },
  }
  return self
}
