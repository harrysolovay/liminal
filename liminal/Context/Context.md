# `Context`

Create a container with a specified key, system message and untouched message
buffer.
