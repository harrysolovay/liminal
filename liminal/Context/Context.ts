import { ActionBase } from "../Action/ActionBase.ts"
import type { ActionLike } from "../Action/ActionLike.ts"
import type { ActorLike } from "../Actor/ActorLike.ts"
import type { ExtractSpec, Spec } from "../Spec.ts"
import type { ContextEnteredEvent, ContextExitedEvent, ContextInnerEvent } from "./ContextEvent.ts"

export interface Context<S extends Spec = Spec> extends ActionBase<"context", S> {
  key: keyof any
  implementation: ActorLike
}

export function* context<K extends keyof any, Y extends ActionLike, S extends ExtractSpec<Y>, T = string>(
  key: K,
  implementation: ActorLike<Y, T>,
): Generator<
  Context<{
    Field: S["Field"]
    Event: ContextEnteredEvent<K> | ContextInnerEvent<K, S["Event"]> | ContextExitedEvent<K, T>
  }>,
  T
> {
  return yield ActionBase("context", {
    key,
    implementation,
  })
}
