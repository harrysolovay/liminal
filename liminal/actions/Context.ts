import type { ActionEvent } from "../ActionEvent.ts"
import type { ActionLike } from "../Actor/ActionLike.ts"
import type { ActorLike } from "../Actor/ActorLike.ts"
import { reduce } from "../Actor/reduce.ts"
import { Scope } from "../Scope.ts"
import type { ExtractSpec, Spec } from "../Spec.ts"
import { unwrapDeferred } from "../util/unwrapDeferred.ts"
import { ActionBase, type ActionEventBase, type EnteredEvent, type ExitedEvent } from "./actions_base.ts"

export interface Context<S extends Spec = Spec> extends ActionBase<"context", S> {
  key: keyof any
  implementation: ActorLike
}

export function* context<K extends keyof any, Y extends ActionLike, S extends ExtractSpec<Y>, T = string>(
  key: K,
  implementation: ActorLike<Y, T>,
): Generator<
  Context<{
    Entry: S["Entry"]
    Event: ContextEvent<K, EnteredEvent | S["Event"] | ExitedEvent<T>>
  }>,
  T
> {
  return yield ActionBase("context", {
    key,
    implementation,
    async reduce(scope) {
      const actor = unwrapDeferred(implementation)
      const events = scope.events.child((event) => ({
        type: "context",
        context: key,
        event,
      }))
      events.emit({
        type: "entered",
      })
      const contextScope = await reduce(
        actor,
        new Scope(
          scope.args,
          key,
          events,
          scope.runInfer,
          scope.runEmbed,
        ),
      )
      events.emit({
        type: "exited",
        result: contextScope.result,
      })
      return scope.spread({
        next: contextScope.result,
        children: [...scope.children, {
          type: "context",
          scope: contextScope,
        }],
      })
    },
  })
}

export interface ContextEvent<K extends keyof any = keyof any, E extends ActionEvent = any>
  extends ActionEventBase<"context">
{
  context: K
  event: E
}
