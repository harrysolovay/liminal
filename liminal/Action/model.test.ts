import { Model } from "./Model.js"
import { Context } from "./Context.js"
import { describe, it, expect } from "bun:test"
import { TestEmbeddingModels } from "../testing/TestEmbeddingModels.js"
import { TestLanguageModels } from "../testing/TestLanguageModels.js"
import type { ActionEvent } from "./ActionEvent.js"

describe("Model", () => {
  it("generates the expected event sequence", async () => {
    const events: Array<ActionEvent> = []
    await Context("Root", function* () {
      yield* Model("default")
      yield* Model("secondary")
      yield* Context("child", function* () {
        yield* Model("child_a")
        yield* Model("child_b", "embedding")
      })
      yield* Model("tertiary", "embedding")
    }).run({
      models: {
        language: TestLanguageModels(),
        embedding: TestEmbeddingModels(),
      },
      handler: (event) => events.push(event),
    })
    expect(events).toMatchSnapshot()
  })
})
