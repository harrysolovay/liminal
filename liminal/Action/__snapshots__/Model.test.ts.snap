// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Model generates the expected event sequence 1`] = `
[
  {
    "event": "Enter",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Enter",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Model",
      "key": "default",
      "purpose": "language",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Model",
      "key": "secondary",
      "purpose": "language",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Context",
      "inner": {
        "event": "Enter",
      },
      "key": "child",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Context",
      "inner": {
        "event": "Model",
        "key": "child_a",
        "purpose": "language",
      },
      "key": "child",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Context",
      "inner": {
        "event": "Model",
        "key": "child_b",
        "purpose": "embedding",
      },
      "key": "child",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Context",
      "inner": {
        "event": "Exit",
        "result": undefined,
      },
      "key": "child",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Model",
      "key": "tertiary",
      "purpose": "embedding",
    },
    "key": "Root",
  },
  {
    "event": "Context",
    "inner": {
      "event": "Exit",
      "result": undefined,
    },
    "key": "Root",
  },
  {
    "event": "Exit",
    "result": undefined,
  },
]
`;
